### PR TITLE
Support javax.inject.Inject as an implicit import

### LINKED ISSUE #13969: Consider adding javax.inject.Inject to default imports
`@Inject` is extensively used in the Gradle programming model.

### Expected Behavior

One can use `@Inject` in Gradle scripts without having to add an import statement.

### Current Behavior

It is required to `import javax.inject.Inject`.
Lots of snippets in documentation need to include the import statement in order to be complete.

### PR DESCRIPTION
**Pull request by bmuskalla**
This allows build authors to use @Inject in their build scripts without explicitly importing the annotation.

Fixes #13969

### CODE REVIEW

File: subprojects/language-java/src/crossVersionTest/groovy/org/gradle/api/internal/tasks/compile/tooling/JavaCompileTaskOperationResultCrossVersionTest.groovy [hunk 1]
```diff
@@ -87,7 +87,7 @@ class JavaCompileTaskOperationResultCrossVersionTest extends ToolingApiSpecifica
     @TargetGradleVersion(">=6.7")
     def "reports annotation processor results for JavaCompile task even when build event listener is used"() {
         settingsFile << """
-            import javax.inject.Inject
+            import javax.inject.Inject;
```
Comment (eskatos) [inline subprojects/language-java/src/crossVersionTest/groovy/org/gradle/api/internal/tasks/compile/tooling/JavaCompileTaskOperationResultCrossVersionTest.groovy:90]: :x: this import statement can be removed now

File: subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/support/ImplicitImports.kt [hunk 2]
```diff
@@ -43,6 +43,7 @@ class ImplicitImports(private val importsReader: ImportsReader) {
             "java.util.concurrent.TimeUnit",
             "java.math.BigDecimal",
             "java.math.BigInteger",
-            "java.io.File"
+            "java.io.File",
+            "javax.inject.Inject"
```
Comment (eskatos) [inline subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/support/ImplicitImports.kt:47]: 👍

File: subprojects/docs/src/snippets/dependencyManagement/artifactTransforms-incremental/groovy/build.gradle [hunk 3]
```diff
@@ -1,11 +1,10 @@
-import javax.inject.Inject
-
 import org.gradle.api.artifacts.transform.TransformParameters
 
 // tag::artifact-transform-countloc[]
 abstract class CountLoc implements TransformAction<TransformParameters.None> {
 
-    @Inject                                                             // <1>
+    @Inject
+    // <1>
```
Comment (eskatos) [inline subprojects/docs/src/snippets/dependencyManagement/artifactTransforms-incremental/groovy/build.gradle:7]: :x: let's keep the callout position untouched

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/suppliers/DynamicRevisionRemoteResolveWithMetadataSupplierIntegrationTest.groovy [hunk 4]
```diff
@@ -1269,7 +1269,7 @@ group:projectB:2.2;release
         file << """import org.gradle.api.artifacts.ComponentMetadataSupplier
           import org.gradle.api.artifacts.ComponentMetadataSupplierDetails
           import org.gradle.api.artifacts.repositories.RepositoryResourceAccessor
-          import javax.inject.Inject
+          import javax.inject.Inject;
```
Comment (eskatos) [inline subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/suppliers/DynamicRevisionRemoteResolveWithMetadataSupplierIntegrationTest.groovy:1272]: :x: this import statement can be removed now

File: subprojects/core/src/main/java/org/gradle/groovy/scripts/internal/GradleResolveVisitor.java [hunk 5]
```diff
@@ -576,7 +577,10 @@ private boolean resolveFromDefaultImports(ClassNode type, boolean testDefaultImp
                     return true;
                 }
             }
-            if (name.equals("BigInteger")) {
+            if (name.equals("Inject")) {
+                type.setRedirect(ClassHelper.makeCached(Inject.class));
+                return true;
+            } else if (name.equals("BigInteger")) {
```
Comment (eskatos) [inline subprojects/core/src/main/java/org/gradle/groovy/scripts/internal/GradleResolveVisitor.java:583]: 👍

**Review by eskatos — CHANGES_REQUESTED** [review]
Great!

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/suppliers/DynamicRevisionRemoteResolveWithMetadataSupplierIntegrationTest.groovy (hunk 4 above)
Comment (bmuskalla) [reply subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/suppliers/DynamicRevisionRemoteResolveWithMetadataSupplierIntegrationTest.groovy:1272]: Thanks for catching

**Comment by bmuskalla** [comment]
@bot-gradle test this

**Comment by bmuskalla** [comment]
@bot-gradle test this

**Comment by bmuskalla** [comment]
@bot-gradle test this

File: subprojects/language-java/src/crossVersionTest/groovy/org/gradle/api/internal/tasks/compile/tooling/JavaCompileTaskOperationResultCrossVersionTest.groovy [hunk 6]
```diff
@@ -87,6 +87,7 @@ class JavaCompileTaskOperationResultCrossVersionTest extends ToolingApiSpecifica
     @TargetGradleVersion(">=6.7")
     def "reports annotation processor results for JavaCompile task even when build event listener is used"() {
         settingsFile << """
+            import javax.inject.Inject
```
Comment (eskatos) [inline subprojects/language-java/src/crossVersionTest/groovy/org/gradle/api/internal/tasks/compile/tooling/JavaCompileTaskOperationResultCrossVersionTest.groovy]: :question: Why is this import required? Default imports should be available in settings scripts

Comment (bmuskalla) [reply subprojects/language-java/src/crossVersionTest/groovy/org/gradle/api/internal/tasks/compile/tooling/JavaCompileTaskOperationResultCrossVersionTest.groovy]: As far as I understand, we need the import because we try to run this test with 6.7 as well and without the implicit import support (6.8), this fails to compile. See[ build failure](https://builds.gradle.org/buildConfiguration/Gradle_Check_Quick_1_bucket19/38744565)

Not sure if we can do something about this

Comment (eskatos) [reply subprojects/language-java/src/crossVersionTest/groovy/org/gradle/api/internal/tasks/compile/tooling/JavaCompileTaskOperationResultCrossVersionTest.groovy]: Ah right. makes sense :+1:

**Review by eskatos — APPROVED** [review]
\o/

**Comment by eskatos** [comment]
This should be mentioned in the release notes
